extensions/csv/src/main/java/io/deephaven/csv/reading/CellGrabber.java

```diff
@@ -0,0 +1,276 @@
+        }
+        if (dest.begin() == dest.end() || dest.back() != quoteChar) {
+            throw new RuntimeException("Logic error: final non-whitespace in field is not quotation mark");
+        }
+        dest.setEnd(dest.end() - 1);
+    }
+
+    private void processUnquotedMode(final CharSlice dest, boolean[] lastInRow) throws CsvReaderException {
+        startOffset = offset;
+        lastInRow[0] = finishField(dest);
+    }
+
+    private void skipWhitespace() throws CsvReaderException {
+        do {
+            while (offset != size) {
+                final char ch = buffer[offset];
+                if (!RangeTests.isWhitespace(ch)) {
+                    return;
+                }
+                ++offset;
+            }
+        } while (tryEnsureMore());
+    }
+
+    /**
+     * @return true iff the field is the last in the row, false otherwise.
+     */
+    private boolean finishField(final CharSlice dest) throws CsvReaderException {
+        while (true) {
+            while (offset != size) {
+                final char ch = buffer[offset];
+                if (ch == fieldDelimiter) {
+                    finish(dest);
+                    ++offset;
+                    return false;
+                }
+                if (ch == '\n') {
+                    finish(dest);
+                    ++offset;
+                    return true;
+                }
+                if (ch == '\r') {
+                    finish(dest);
+                    ++offset;
+                    if (tryEnsureMore()) {
+                        // might be \r\n
+                        if (buffer[offset] == '\n') {
+                            ++offset;
+                        }
+                    }
+                    return true;
+                }
+                ++offset;
+            }
+            if (!tryEnsureMore()) {
+                finish(dest);
+                // End of file sets last in row.
+                return true;
+            }
+        }
+    }
+
+    /**
+     * @return true if there are more characters.
+     */
+    private boolean tryEnsureMore() throws CsvReaderException {
+        if (offset != size) {
+            return true;
+        }
+        spillRange();
+        refillBuffer();
+        return size != 0;
+    }
+
+    private void spillRange() {
+        spillBuffer.append(buffer, startOffset, offset - startOffset);
+        startOffset = offset;
+    }
+
+    private void refillBuffer() throws CsvReaderException {
+        offset = 0;
+        startOffset = 0;
+        try {
+            while (true) {
+                final int charsRead = reader.read(buffer);
+                if (charsRead < 0) {
+                    size = 0;
+                    return;
+                }
+                if (charsRead > 0) {
+                    size = charsRead;
+                    return;
+                }
+                // I don't think zero-length reads happen, but if they do, I'll just keep trying.
```

Comment:
I suggest rewording to avoid first person singular, eg:
"Normally zero-length reads shouldn't happen; if they do, we just keep trying."

Reply:
Changed in a different way. thanks.
```
// Zero-length reads are not expected to happen, but if they do, just keep trying.
```